From 46ca41e55a35bf2022731b0f3cf119e6f51d984c Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Fri, 21 Jul 2023 11:43:10 +0800
Subject: [PATCH 11/53] [Huawei] mm/sharepool: Delete state element in
 sp_alloc_context

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

The state element is used for hugepage-fallback and coredump-check. Both
of them can be replaced by the return value from the callee.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 45 ++++++++++++++-------------------------------
 1 file changed, 14 insertions(+), 31 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index c5a712d73ff9..468f2753fedb 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -2103,17 +2103,11 @@ static unsigned long sp_mmap(struct mm_struct *mm, struct file *file,
 	return addr;
 }
 
-#define ALLOC_NORMAL	1
-#define ALLOC_RETRY	2
-#define ALLOC_NOMEM	3
-#define ALLOC_COREDUMP	4
-
 struct sp_alloc_context {
 	struct sp_group *spg;
 	unsigned long size;
 	unsigned long size_aligned;
 	unsigned long sp_flags;
-	int state;
 	bool have_mbind;
 	enum spa_type type;
 };
@@ -2179,26 +2173,29 @@ static int sp_alloc_prepare(unsigned long size, unsigned long sp_flags,
 	ac->spg = spg;
 	ac->size = size;
 	ac->sp_flags = sp_flags;
-	ac->state = ALLOC_NORMAL;
 	ac->have_mbind = false;
 	return 0;
 }
 
-static void sp_alloc_fallback(struct sp_area *spa, struct sp_alloc_context *ac)
+static bool sp_alloc_fallback(struct sp_area *spa, struct sp_alloc_context *ac)
 {
 	/*
 	 * If hugepage allocation fails, this will transfer to normal page
 	 * and try again. (only if SP_HUGEPAGE_ONLY is not flagged
 	 */
-	if (!(ac->sp_flags & SP_HUGEPAGE) || (ac->sp_flags & SP_HUGEPAGE_ONLY)) {
-		ac->state = ALLOC_NOMEM;
-		return;
-	}
+	if (!(ac->sp_flags & SP_HUGEPAGE) || (ac->sp_flags & SP_HUGEPAGE_ONLY))
+		return false;
 
 	ac->size_aligned = ALIGN(ac->size, PAGE_SIZE);
 	ac->sp_flags &= ~SP_HUGEPAGE;
-	ac->state = ALLOC_RETRY;
+	/*
+	 * The mempolicy for shared memory is located at backend file, which varies
+	 * between normal pages and huge pages. So we should set the mbind policy again
+	 * when we retry using normal pages.
+	 */
+	ac->have_mbind = false;
 	__sp_area_drop(spa);
+	return true;
 }
 
 static long sp_mbind(struct mm_struct *mm, unsigned long start, unsigned long len,
@@ -2208,6 +2205,7 @@ static long sp_mbind(struct mm_struct *mm, unsigned long start, unsigned long le
 			nodemask, MPOL_MF_STRICT, mm);
 }
 
+#define SP_SKIP_ERR 1
 static int __sp_alloc_mmap_populate(struct mm_struct *mm, struct sp_area *spa,
 	struct sp_group_node *spg_node, struct sp_alloc_context *ac,
 	nodemask_t *nodemask)
@@ -2222,9 +2220,8 @@ static int __sp_alloc_mmap_populate(struct mm_struct *mm, struct sp_area *spa,
 	down_write(&mm->mmap_lock);
 	if (unlikely(mm->core_state)) {
 		up_write(&mm->mmap_lock);
-		ac->state = ALLOC_COREDUMP;
 		pr_info("allocation encountered coredump\n");
-		return -EFAULT;
+		return SP_SKIP_ERR;
 	}
 
 	if (spg_node)
@@ -2303,17 +2300,14 @@ static int sp_alloc_mmap_populate(struct sp_area *spa,
 		mm = spg_node->master->mm;
 		mmap_ret = __sp_alloc_mmap_populate(mm, spa, spg_node, ac, &__nodemask);
 		if (mmap_ret) {
-
 			/*
 			 * Goto fallback procedure upon ERR_VALUE,
 			 * but skip the coredump situation,
 			 * because we don't want one misbehaving process to affect others.
 			 */
-			if (ac->state != ALLOC_COREDUMP)
+			if (mmap_ret != SP_SKIP_ERR)
 				goto unmap;
 
-			/* Reset state and discard the coredump error. */
-			ac->state = ALLOC_NORMAL;
 			continue;
 		}
 		ret = mmap_ret;
@@ -2335,10 +2329,6 @@ static int sp_alloc_mmap_populate(struct sp_area *spa,
 	 */
 	sp_fallocate(spa);
 
-	/* if hugepage allocation fails, this will transfer to normal page
-	 * and try again. (only if SP_HUGEPAGE_ONLY is not flagged */
-	sp_alloc_fallback(spa, ac);
-
 	return mmap_ret;
 }
 
@@ -2386,15 +2376,8 @@ void *__mg_sp_alloc_nodemask(unsigned long size, unsigned long sp_flags, int spg
 	}
 
 	ret = sp_alloc_mmap_populate(spa, &ac, nodemask);
-	if (ret && ac.state == ALLOC_RETRY) {
-		/*
-		 * The mempolicy for shared memory is located at backend file, which varies
-		 * between normal pages and huge pages. So we should set the mbind policy again
-		 * when we retry using normal pages.
-		 */
-		ac.have_mbind = false;
+	if (ret == -ENOMEM && sp_alloc_fallback(spa, &ac))
 		goto try_again;
-	}
 
 out:
 	sp_alloc_finish(ret, spa, &ac);
-- 
2.17.1

